### PR TITLE
xbps-src: treat XBPS_FETCH_CMD as a user-definable variable

### LINKED ISSUE #3399: XBPS_FETCH_CMD leakage
### System

* xuname:  Void 4.18.10_1 x86_64 GenuineIntel notuptodate rrmFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  *output of ``xuname`` (part of xtools)*
* package: xbps-src 
  *affected package(s) including the version*

### Expected behavior

All packages compile successfully

### Actual behavior

XBPS_FETCH_CMD leaks into another package build

### Steps to reproduce the behavior

1. rm -rf hostdir/sources/perl-Curses
2. add perl-Curses to hostmakedepends= of `asciiquarium`
3. xrevbump 'bump' perl-Curses
4. ./xbps-src build asciiquarium

xbps-src sees perl-Curses needs to be rebuilt and goes fetch the source code but since it already sourced `asciiquarium` template it uses the XBPS_FETCH_CMD from it which uses wget.

```
=> asciiquarium-1.1_3: building ...
   [host] wget-1.19.5_1: found (https://alpha.de.repo.voidlinux.org/current)
   [host] perl-Curses-1.36_2: not found.
   [runtime] perl>=0: found perl-5.28.0_2 (https://alpha.de.repo.voidlinux.org/current)
   [runtime] perl-Term-Animation>=0: found perl-Term-Animation-2.6_2 (https://alpha.de.repo.voidlinux.org/current)
   [runtime] perl-Curses>=0: found perl-Curses-1.36_1 (https://alpha.de.repo.voidlinux.org/current)
=> perl-Curses-1.36_2: building [perl-module] (dependency of asciiquarium) ...
   [host] perl-5.28.0_2: found (https://alpha.de.repo.voidlinux.org/current)
   [target] perl-5.28.0_2: found (https://alpha.de.repo.voidlinux.org/current)
   [target] ncurses-devel-6.1_2: found (https://alpha.de.repo.voidlinux.org/current)
=> perl-Curses: building [perl-module] (dependency of asciiquarium) ...
=> perl-Curses-1.36_2: installing host dependency 'perl-5.28.0_2' ...
=> perl-Curses-1.36_2: installing target dependency 'perl-5.28.0_2' ...
=> perl-Curses-1.36_2: installing target dependency 'ncurses-devel-6.1_2' ...
=> perl-Curses-1.36_2: running do-fetch hook: 00-distfiles ...
=> perl-Curses-1.36_2: fetching distfile 'Curses-1.36.tar.gz'...
flock: failed to execute wget: No such file or directory
=> ERROR: perl-Curses-1.36_2: failed to fetch Curses-1.36.tar.gz.
```

### LINKED ISSUE #7513: XBPS_FETCH_CMD is poorly treated
XBPS_FETCH_CMD allows users to modify which program to fetch distfiles= and is by default `xbps-uhelper fetch`.

some templates modify this variable but xbps-src itself doesn't actually treat it as a modifiable variable.

- It is setup on `xbps-src`, **once**. If you modify it, it will propagate to any packages built in a single xbps-src call.
- it isn't unset on common/environment/setup/sourcepkg.sh, so it leaks if modified to other templates, and it can't be unset because the variable is set on `xbps-src` **once**, unsetting it causes it to become undefined.

### PR DESCRIPTION
**Pull request by maxice8**
closes #7513
closes #3399

### CODE REVIEW

**Comment by Duncaen** [comment]
Wondering if its better if all `XBPS_` variables are read only a new template variable like `fetch_cmd` is introduced.

Also libfetch should be able to manage all the cases where wget is currently used, except maybe the dropbox download, not sure what the issue is there.
For setting the referer libfetch has `HTTP_REFERER` and `SSL_NO_VERIFY_PEER` or some other more specific environment variable for disabling certificate checks.

File: common/xbps-src/shutils/common.sh [hunk 1]
```diff
@@ -592,6 +592,12 @@ setup_pkg() {
     export XBPS_STATEDIR="${XBPS_BUILDDIR}/.xbps-${sourcepkg}"
     export XBPS_WRAPPERDIR="${XBPS_STATEDIR}/wrappers"
 
+    # Set default XBPS_FETCH_CMD in case the template doesn't
+    # define it.
+    if [ -z "$XBPS_FETCH_CMD" ]; then
+        export XBPS_FETCH_CMD="xbps-uhelper fetch"
+    fi
```
Comment (Gottox) [inline common/xbps-src/shutils/common.sh]: The variable should always be exported, not only when it's not set:
```
    export XBPS_FETCH_CMD=${XBPS_FETCH_CMD:-xbps-uhelper fetch}
```

**Comment by maxice8** [comment]
On Mon, Jan 21, 2019 at 03:59:16AM -0800, Duncan Overbruck wrote:
> Wondering if its better if all `XBPS_` variables are read only a new template variable like `fetch_cmd` is introduced.
> 
> Also libfetch should be able to manage all the cases where wget is currently used, except maybe the dropbox download, not sure what the issue is there.
> For setting the referer libfetch has `HTTP_REFERER` and `SSL_NO_VERIFY_PEER` or some other more specific environment variable for disabling certificate checks.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/pull/7514#issuecomment-456050166

I'm in favour of adding fetch_cmd, making XBPS_FETCH_CMD readonly like all other XBPS_ and switching the templates over.

**Comment by Gottox** [comment]
or refactor the affected templates to use do_fetch instead of overwriting XBPS_FETCH_CMD.

**Comment by maxice8** [comment]
Pushed fetch_cmd solution, please check

**Pull request merged**